### PR TITLE
Fix store cleanup race condition

### LINKED ISSUE #193: Container cleanup occurs after remounting, clearing state set by onInit
**Scenario:** 

I have a component that is linked with a container and has a scope.

**Events:** 

* mount the component
* onInit is fired and sets some state
* unmount the component - triggers the state cleanup to occur, but does not execute immediately
* mount the component (same component)
* onInit is fired and sets some state, the state from the previous render is still present, component sets some state
* cleanup occurs, removing the state

This situation happens within react storybooks. I tried to reproduce it in pure react using two different components linked to the same container to ensure react performs an unmount and remount when switching, but did not experience the same problem. Perhaps I'm missing something here though.

I did notice that the cleanup happens on the next tick, so in theory it should be possible to replicate it in other environments.
https://github.com/atlassian/react-sweet-state/blob/master/src/components/container.js#L79

**Expected:**

I would expect the cleanup to occur before mounting componentB. This option is preferred as it ensures you get clean state if you unmount the last component using a store, then remount another component using the same store.

Alternatively not perform the cleanup step at all if it would trigger after the same container is remounted. This would at least ensure the state is not destroyed, but state would be persisted across components even though the store was unmounted from the application entirely.

**Example:**

In my specific case, this causes a problem with using a singular component in storybooks.

https://codesandbox.io/p/sandbox/hungry-sea-ub8boh?file=%2Fsrc%2Ftest-components.story.tsx&selection=%5B%7B%22endColumn%22%3A25%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A25%2C%22startLineNumber%22%3A4%7D%5D

To see the issue, switch between the stories and view the logs.

### PR DESCRIPTION
**Pull request by albertogasparin**
Due to the async nature of cleanup, we ensure registry `deleteStore` is called only if the Container store instance has not been replaced in meantime.

Closes #193

### CODE REVIEW

File: src/components/container.js [hunk 1]
```diff
@@ -151,22 +157,15 @@ export default class Container extends Component {
...
+  deleteScopedStore(prevStoreState, scopeId = this.props.scope) {
+    const { storeState } = this.getScopedStore(scopeId);
+    if (
+      scopeId != null &&
+      !prevStoreState.listeners().length &&
+      // ensure registry has not already created a new store w/ same scope
+      prevStoreState === storeState
+    ) {
+      const { storeType } = this.constructor;
       this.getRegistry().deleteStore(storeType, scopeId);
```
Comment (theKashey) [inline src/components/container.js:169]: I have a strong feeling that the guard rail should be in `deleteStore`, not React component.

File: src/components/__tests__/container.test.js [hunk 2]
```diff
@@ -194,6 +195,23 @@ describe('Container', () => {
...
 
+    it('should not cleanup from global if same type gets recreated meanwhile', async () => {
+      const Subscriber = createSubscriber(Store);
+      const renderPropChildren = jest.fn().mockReturnValue(null);
+      const children = <Subscriber>{renderPropChildren}</Subscriber>;
+      const { unmount } = render(<Container scope="s1">{children}</Container>);
+      unmount();
+
+      // given cleanup is scheduled, we fake the eventuality that the store gets replaced in meantime
+      defaultRegistry.getStore.mockReturnValue({
```
Comment (theKashey) [inline src/components/__tests__/container.test.js:206]: so this is to _support_ `prevStoreState === storeState`, but it's sooooo synthetic.

Can the same example be recreated by flipping `scope` on sibling components, so they will change it in one tick?

Comment (albertogasparin) [reply src/components/__tests__/container.test.js:206]: Tried that in multiple ways, but cannot replicate in test environment. React has lots of special cases around how `useEffect` are triggered and could not find a way to have the cleanup happen between render and `useEffect` in tests 😞 
One alternative is to mock `useExternalStore`, which feels as dirty as this test thought

File: src/components/container.js (hunk 1 above)
Comment (albertogasparin) [reply src/components/container.js:169]: Maybe. Currently the registry does not keep track of the stores lifecycle, so if we ask to delete something why should it say no? 
We could shift that management to the Registry itself, but would need to implement some signals/comms layer to let registry know that store subscribers have changed, that a container has been unmounted, ... 
It could be a more solid solution, but I feel it might be too elaborate. At the end of the day this is an edge case, and maybe we can handle it better when we will rewrite this class into a functional component.

File: src/components/__tests__/container.test.js (hunk 2 above)
Comment (theKashey) [reply src/components/__tests__/container.test.js:206]: 🤗 I feel you. The condition you want to represent synthetically seems to be equal to the reported problem. It will be good to have a "real use case", but if it's too hard to reproduce in _another environment_ - one can always verify manually, and if the problem is solved, then the test is doing the right thing and will protect behavior from future regressions.

**Review by theKashey — APPROVED** [review]
